perf: Implement filtering in search page

TMisiukiewicz opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->
### Details
<!-- Explanation of the change or anything fishy that is going on -->
Currently when using search input, the option list is generated from scratch. With this change, list is created only once and it is filtered when value of search input change. Once we implement this in all search pages, we will be able to remove `searchText` property from options, which will have positive impact on execution time of `getOptions` (around 25% improvement). From our measurements, filtering is ~12x faster than creating an option list from scratch.
### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/37619
PROPOSAL:


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->
1. Open Search Page
2. Type something in the search input
3. Verify results on the list are filtered to match the searched phase
4. Compare the results and verify they are displayed in the same order as on production app
5. Clear the input and type some random characters
6. Verify "No results found" message is visible
- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->
1. Open Search Page
2. Type something in the search input
3. Verify results on the list are filtered to match the searched phase
4. Compare the results and verify they are displayed in the same order as on production app
5. Clear the input and type some random characters
6. Verify "No results found" message is visible

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
7. Upload an image via copy paste
8. Verify a modal appears displaying a preview of that image
--->
1. Open Search Page
2. Type something in the search input
3. Verify results on the list are filtered to match the searched phase
4. Compare the results and verify they are displayed in the same order as on production app
5. Clear the input and type some random characters
6. Verify "No results found" message is visible
- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/13985840/82648d3e-cec6-42e9-9806-a2d83d9242ca


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/13985840/6d04d827-eb9a-48ec-98df-1b0380963277


<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/13985840/78b22fbf-17a4-43b8-b667-1158f5e9ef55


<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/13985840/9c317d4a-ec60-4e11-8575-68c16a120707


<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/13985840/01949f56-41f6-4430-b78a-a3eecedeca31


<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/13985840/307188a4-40d8-434f-87bc-baae4cf3e8e1


</details>

TMisiukiewicz committed:
incorporate match-based filtering into codebase

TMisiukiewicz committed:
create function responsible for filtering options

TMisiukiewicz committed:
use filtering on search page

TMisiukiewicz committed:
adapt alghoritm file to match project standards

TMisiukiewicz committed:
update filtering method & refactor options filtering

TMisiukiewicz committed:
remove comments

TMisiukiewicz committed:
update search page to filter options

TMisiukiewicz committed:
add regex email to filtering results

TMisiukiewicz committed:
remove sorting

jbroma commented on src/libs/filterArrayByMatch.ts:
would be nice to leave a comment that this is a slim version of what's available in `match-sorter`

gedu commented on src/libs/OptionsListUtils.ts:
```diff
@@ -1997,6 +2011,73 @@ function formatSectionsFromSearchTerm(
+
+    // The regex below is used to remove dots only from the local part of the user email (local-part@domain)
+    // so that we can match emails that have dots without explicitly writing the dots (e.g: fistlast@domain will match first.last@domain)
+    const emailRegex = /\.(?=[^\s@]*@)/g;
```
probably you can initialize it outside of the function

TMisiukiewicz replied:
hmm, I think it might negatively affect the performance, because we would have to call it for each value under a specified keys. With current approach, we can limit those checks only to the keys we need.

gedu commented on src/libs/filterArrayByMatch.ts:
```diff
@@ -0,0 +1,326 @@
+    } else if (keyCopy.includes('.')) {
+        return getNestedValues<T>(keyCopy, item);
+    } else {
+        value = null;
```
probably you can do return []; here, instead of assign null and then check for null and return

TMisiukiewicz committed:
update filter by match

TMisiukiewicz committed:
update filtered options state

TMisiukiewicz committed:
Merge branch 'main' into perf/search-options-filtering

TMisiukiewicz committed:
reduce amount of rerenders

TMisiukiewicz committed:
update linting

TMisiukiewicz committed:
Merge branch 'main' into perf/search-options-filtering

TMisiukiewicz committed:
Merge remote-tracking branch 'origin/main' into perf/search-options-filtering

TMisiukiewicz committed:
filter by phone number

TMisiukiewicz committed:
update search options header

TMisiukiewicz committed:
revert package lock

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/search-options-filtering

TMisiukiewicz committed:
fix package-lock

DylanDylann reviewed (commented):
I will continue reviewing this PR early tomorrow

DylanDylann commented on src/libs/OptionsListUtils.ts:
```diff
@@ -2032,6 +2046,75 @@ function formatSectionsFromSearchTerm(
 
+function filterOptions(options: Partial<GetOptions>, searchInputValue: string): ReportUtils.OptionData[] {
+    const parsedPhoneNumber = PhoneNumber.parsePhoneNumber(LoginUtils.appendCountryCode(Str.removeSMSDomain(searchInputValue)));
+    const searchValue = parsedPhoneNumber.possible ? parsedPhoneNumber.number?.e164 : searchInputValue.toLowerCase();
```
`searchValue` is defined two times. Let's DRY code

TMisiukiewicz committed:
Merge branch 'main' into perf/search-options-filtering

DylanDylann commented:
@TMisiukiewicz Please help to merge the latest main

TMisiukiewicz commented on src/pages/SearchPage/index.tsx:
```diff
@@ -88,10 +84,32 @@ function SearchPage({betas, reports, isSearchingForReports, navigation}: SearchP
+    }, [betas, isScreenTransitionEnd, personalDetails, reports]);
+
+    const filteredOptions = useMemo(() => {
+        if (debouncedSearchValue.trim() === '') {
```
it is used to prevent unnecessarily calling the `filterOptions` function when component is re-rendering, I'd say we should keep it here

DylanDylann replied:
Whoops, `filteredOptions` is a variable, not a function. Look good

DylanDylann commented:
@TMisiukiewicz I see that there are some differences in the search result after applying your change
 
https://github.com/Expensify/App/assets/141406735/509bdc55-1802-4497-8ed2-a9f08f472169

@mountiny Could we ping a QA engineer to verify if this change is acceptable?

TMisiukiewicz committed:
wrap parsing phone number in a function

mountiny commented:
@DylanDylann @TMisiukiewicz Hmm, that makes me feel like there should be unit tests for the search order. We could add them in this PR. 

Could we define the difference in the order based on the logic changes?

TMisiukiewicz committed:
fix tests

mountiny commented:
What exactly is the difference in the order based on the logic? 

I think ideally we have unit tests covering this all. @TMisiukiewicz I believe this change should not influence the order so would you be able to look into these findings and ideally - add unit tests for the search order on main (separate PR) such that we confirm the order is unchanged.

There are already tests for the order in the LHN, but possibly missing for the search list

TMisiukiewicz commented:
I think the main difference in this approach is the fact that i am filtering reports, personal details and chat rooms/policy expense chats separately, because they are filtered by different keys. I'll try to adjust the solution to return it in unchanged order. Thanks for spotting this!
@mountiny do you want to get tests merged into main first, before this PR?

mountiny commented:
Yep ideally we can merge the tests into main first

TMisiukiewicz committed:
rewrite filtering function

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/search-options-filtering

TMisiukiewicz committed:
update phone parsing function

TMisiukiewicz commented:
I have refactored the way of filtering to better reflect the current flow. I was testing it on three different accounts with multiple combinations and each time I got the same results, which makes me think it works fine. Tomorrow morning I'll just double-check the performance impact of this change.
@DylanDylann could you test it once again? Ideally compare the results from this branch to the results you get on main.
@mountiny I'm gonna work on the tests tomorrow 👍 

Once we get this one merged, I'll continue creating PRs for other lists that are using `getOptions`, so once we migrate all of them, we will be able to remove `searchText` param from options, which will speed up search opening

TMisiukiewicz committed:
option list utils cleanup

TMisiukiewicz committed:
add tests for filtering

TMisiukiewicz committed:
add test checking lastVisibleActionCreated

TMisiukiewicz commented:
@mountiny I have added some unit tests directly in this PR, because existing mechanism is already partially covered with tests. To avoid creating a separate PR and then overwriting all the function occurencies here, I was testing the original function first, and then expecting the same output from the new filtering method, I hope this will help us move faster with this feature

mountiny commented:
Thank you, @DylanDylann  are you able to continue the review now?

DylanDylann commented on src/libs/OptionsListUtils.ts:
```diff
@@ -1438,8 +1474,7 @@ function getOptions(
     const reportMapForAccountIDs: Record<number, Report> = {};
-    const parsedPhoneNumber = PhoneNumber.parsePhoneNumber(LoginUtils.appendCountryCode(Str.removeSMSDomain(searchInputValue)));
-    const searchValue = parsedPhoneNumber.possible ? parsedPhoneNumber.number?.e164 : searchInputValue.toLowerCase();
+    const {output: searchValue, parsedPhoneNumber} = getSearchValueForPhoneOrEmail(searchInputValue);
```
I don't think adding parsedPhoneNumber to the result of `getSearchValueForPhoneOrEmail` function is good. Because we only use `parsedPhoneNumber` in here. Let's keep `getSearchValueForPhoneOrEmail` function as before and use it here
```
const parsedPhoneNumber = PhoneNumber.parsePhoneNumber(LoginUtils.appendCountryCode(Str.removeSMSDomain(searchInputValue)));
const {output: searchValue, parsedPhoneNumber} = getSearchValueForPhoneOrEmail(searchInputValue);
```

TMisiukiewicz replied:
I just decided to revert this change, because parsePhoneNumber might have negative impact on the performance (it's doing operations based on regexes under the hood). Once all the lists migrate to this type of search, all the code related to search from this function could be possibly removed, including this one, so I think we could leave it unchanged for now and remove in the future, to prevent excessive function calls

DylanDylann commented:
Everything else looks good, I am testing

TMisiukiewicz committed:
Merge branch 'main' into perf/search-options-filtering

DylanDylann commented:
More information: Previously we had UTs for the search feature

https://github.com/Expensify/App/blob/96716fbf18d7179289eaf59a53928d0727f849ac/tests/unit/OptionsListUtilsTest.js#L314

DylanDylann commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

DylanDylann commented:
@TMisiukiewicz In the test steps, we need to have a step to verify that the result are the same as the old implementation

TMisiukiewicz committed:
Revert "update phone parsing function"

This reverts commit f370b0299a8b487443c3feb553363d7fff9f97fa.

TMisiukiewicz commented:
@DylanDylann thanks, I just updated the test plan with an additional step

DylanDylann commented:
When testing I see that the result is the same as the latest main but different from the staging

DylanDylann commented:
> Compare the results and verify they are displayed in the same order as on production app

Maybe the result is the same as the old logic but different from production

DylanDylann commented:
Failed perf comes from the latest main

DylanDylann commented:
@roryabraham All yours

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/search-options-filtering

roryabraham commented on src/libs/filterArrayByMatch.ts:
```diff
@@ -0,0 +1,327 @@
+ * @returns the highest ranking
+ */
+function getHighestRanking<T>(item: T, keys: ReadonlyArray<KeyOption<T>> | undefined, value: string, options: Options<T>): RankingInfo {
+    if (!keys) {
```
in our case `keys` is always provided, so I think we can make it required and get rid of this base case rather than building things we don't use or intend to use

roryabraham commented on src/libs/filterArrayByMatch.ts:
```diff
@@ -0,0 +1,327 @@
+    }
+
+    // will return a number between rankings.MATCHES and rankings.MATCHES + 1 depending  on how close of a match it is.
+    return getClosenessRanking(lowercaseTestString, lowercaseStringToRank);
```
the difference between `getMatchRanking` and `getClosenessRanking` ranking is not immediately clear. Given that `getClosenessRanking` is only used once, let's just inline it here.

roryabraham reviewed (changes requested):
My high-level feedback is that this is much more complex than it needs to be. For context, it seems like you tried to rebuild match-sorter 1:1, but one of the main reasons we shy away from 3rd party dependencies is because they often include complex code and features that we don't necessarily need. Please review the PR and ensure that:

- we aren't building anything we don't need
- we keep abstractions and single-use wrapper functions to a reasonable minimum
- try to remove redundant config options to simplify types and keep things more consistent

TMisiukiewicz committed:
Merge branch 'main' into perf/search-options-filtering

TMisiukiewicz committed:
code review updates

TMisiukiewicz committed:
move getAcronym to string utils

TMisiukiewicz commented on src/libs/filterArrayByMatch.ts:
```diff
@@ -0,0 +1,327 @@
+ * @param path a dot-separated set of keys
+ * @param item the item to get the value from
+ */
+function getNestedValues<T>(path: string, item: T): string[] {
```
there is also a case with a `*` wildcard, which allows to filter arrays of multiple nested objects. I feel like it's a good part of the api, as it does not require to write this logic manually each time

roryabraham replied:
we aren't using wildcard matching anywhere, so let's please remove this for simplicity

TMisiukiewicz committed:
remove createFilter function

TMisiukiewicz committed:
update reduce with array function

TMisiukiewicz committed:
remove getClosenessRanking function

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/search-options-filtering

TMisiukiewicz committed:
remove keys check

TMisiukiewicz committed:
update tests

roryabraham commented on src/libs/OptionsListUtils.ts:
```diff
@@ -1402,6 +1403,35 @@ function createOptionFromReport(report: Report, personalDetails: OnyxEntry<Perso
+ * @param searchValue - search string
+ * @returns a sorted list of options
+ */
+function orderOptions(options: ReportUtils.OptionData[], searchValue: string | undefined) {
```
let's add some unit tests to cover this function?

TMisiukiewicz replied:
tests for ordering are already covered with `filterOptions` tests

roryabraham commented on src/libs/OptionsListUtils.ts:
```diff
@@ -2086,6 +2097,96 @@ function formatSectionsFromSearchTerm(
+/**
+ * Filters options based on the search input value
+ */
+function filterOptions(options: GetOptions, searchInputValue: string): GetOptions {
```
I know this wasn't you, but `GetOptions` is a very strange name for this type

roryabraham commented on src/libs/OptionsListUtils.ts:
```diff
@@ -2086,6 +2097,96 @@ function formatSectionsFromSearchTerm(
+        return keys;
+    };
+
+    const matchResults = searchTerms.reduceRight((items, term) => {
```
I'm a bit confused why we're doing a `reduce` here, is it intentional that we call `filterArrayByMatch` separately for every search term? doesn't think mean we're filtering over all the results iteratively, once for each search term? Wouldn't it be more efficient to filter all the results once, accounting for all the search terms in a single pass?

TMisiukiewicz committed:
change GetOptions type to Options

TMisiukiewicz committed:
remove dot-notation pattern

TMisiukiewicz committed:
Merge remote-tracking branch 'upstream/main' into perf/search-options-filtering

TMisiukiewicz committed:
Merge branch 'main' into perf/search-options-filtering

TMisiukiewicz committed:
simplify the implementation

TMisiukiewicz committed:
Merge branch 'main' into perf/search-options-filtering

TMisiukiewicz committed:
Merge branch 'main' into perf/search-options-filtering

TMisiukiewicz committed:
Merge branch 'main' into perf/search-options-filtering

TMisiukiewicz committed:
fix typecheck

TMisiukiewicz committed:
remove debounced state dependency

TMisiukiewicz commented:
@roryabraham I've addressed all of your comments. I have reduced the filtering implementation by another ~50 LOC. Looking forward to get your opinion on current state of the PR

TMisiukiewicz commented:
I think there is some issue with reassure job, could anyone re-run it?

roryabraham commented on src/libs/filterArrayByMatch.ts:
```diff
@@ -0,0 +1,195 @@
+type IndexableByString = Record<string, unknown>;
+
+/**
+ * Gets value for key in item at arbitrarily nested keypath
```
This comment description doesn't seem correct anymore?

roryabraham reviewed (changes requested):
As we iterate on this, more simplifications are becoming apparent:

- we aren't actually using different match thresholds at all
- we certainly aren't using different match thresholds on a per-key basis
- all "keys" are currently functions, but typed to be keys or functions that return either a string or array of string

Here's a diff that accomplishes the same thing we have here, more simply:

```diff
diff --git a/src/libs/OptionsListUtils.ts b/src/libs/OptionsListUtils.ts
index 33820da834..530e7e2b6a 100644
--- a/src/libs/OptionsListUtils.ts
+++ b/src/libs/OptionsListUtils.ts
@@ -2246,39 +2246,35 @@ function filterOptions(options: Options, searchInputValue: string): Options {
         return keys;
     };
     const matchResults = searchTerms.reduceRight((items, term) => {
-        const recentReports = filterArrayByMatch(items.recentReports, term, {
-            keys: [
-                (item) => {
-                    let keys: string[] = [];
-                    if (item.text) {
-                        keys.push(item.text);
-                    }
+        const recentReports = filterArrayByMatch(items.recentReports, term, (item) => {
+            let values: string[] = [];
+            if (item.text) {
+                values.push(item.text);
+            }
 
-                    if (item.login) {
-                        keys.push(item.login);
-                        keys.push(item.login.replace(emailRegex, ''));
-                    }
+            if (item.login) {
+                values.push(item.login);
+                values.push(item.login.replace(emailRegex, ''));
+            }
 
-                    if (item.isThread) {
-                        if (item.alternateText) {
-                            keys.push(item.alternateText);
-                        }
-                        keys = keys.concat(getParticipantsLoginsArray(item));
-                    } else if (!!item.isChatRoom || !!item.isPolicyExpenseChat) {
-                        if (item.subtitle) {
-                            keys.push(item.subtitle);
-                        }
-                    } else {
-                        keys = keys.concat(getParticipantsLoginsArray(item));
-                    }
+            if (item.isThread) {
+                if (item.alternateText) {
+                    values.push(item.alternateText);
+                }
+                values = values.concat(getParticipantsLoginsArray(item));
+            } else if (!!item.isChatRoom || !!item.isPolicyExpenseChat) {
+                if (item.subtitle) {
+                    values.push(item.subtitle);
+                }
+            } else {
+                values = values.concat(getParticipantsLoginsArray(item));
+            }
 
-                    return uniqFast(keys);
-                },
-            ],
-        });
-        const personalDetails = filterArrayByMatch(items.personalDetails, term, {
-            keys: [(item) => item.participantsList?.[0]?.displayName ?? '', 'login', (item) => item.login?.replace(emailRegex, '') ?? ''],
+            return uniqFast(values);
         });
+        const personalDetails = filterArrayByMatch(items.personalDetails, term, (item) =>
+            uniqFast([item.participantsList?.[0]?.displayName ?? '', item.login?.replace(emailRegex, '') ?? '']),
+        );
 
         return {
             recentReports: recentReports ?? [],
diff --git a/src/libs/filterArrayByMatch.ts b/src/libs/filterArrayByMatch.ts
index f481a08fe0..c87ad8fac1 100644
--- a/src/libs/filterArrayByMatch.ts
+++ b/src/libs/filterArrayByMatch.ts
@@ -18,59 +18,6 @@ const MATCH_RANK = {
 
 type Ranking = ValueOf<typeof MATCH_RANK>;
 
-type RankingInfo = {
-    rankedValue: string;
-    rank: Ranking;
-    keyIndex: number;
-    keyThreshold?: Ranking;
-};
-
-type ValueGetterKey<T> = (item: T) => string | string[];
-
-type KeyAttributesOptions<T> = {
-    key: string | ValueGetterKey<T>;
-    threshold?: Ranking;
-};
-
-type KeyOption<T> = KeyAttributesOptions<T> | ValueGetterKey<T> | string;
-
-type Options<T = unknown> = {
-    keys: ReadonlyArray<KeyOption<T>>;
-    threshold?: Ranking;
-};
-type IndexableByString = Record<string, unknown>;
-
-/**
- * Gets value for key in item at arbitrarily nested keypath
- * @param item - the item
- * @param key - the potentially nested keypath or property callback
- * @returns an array containing the value(s) at the nested keypath
- */
-function getItemValues<T>(item: T, key: KeyOption<T>): string[] {
-    if (!item) {
-        return [];
-    }
-
-    const resolvedKey = typeof key === 'object' ? key.key : key;
-    const value = typeof resolvedKey === 'function' ? resolvedKey(item) : (item as IndexableByString)[resolvedKey];
-
-    if (!value) {
-        return [];
-    }
-
-    return Array.isArray(value) ? value.map(String) : [String(value)];
-}
-
-/**
- * Gets all the values for the given keys in the given item and returns an array of those values
- * @param item - the item from which the values will be retrieved
- * @param keys - the keys to use to retrieve the values
- * @return objects with {itemValue}
- */
-function getAllValuesToRank<T>(item: T, keys: ReadonlyArray<KeyOption<T>>): string[] {
-    return keys.flatMap((key) => getItemValues(item, key));
-}
-
 /**
  * Gives a rankings score based on how well the two strings match.
  * @param testString - the string to test against
@@ -140,56 +87,31 @@ function getMatchRanking(testString: string, stringToRank: string): Ranking {
     return ranking as Ranking;
 }
 
-/**
- * Gets the highest ranking for value for the given item based on its values for the given keys
- * @param item - the item to rank
- * @param keys - the keys to get values from the item for the ranking
- * @param value - the value to rank against
- * @param options - options to control the ranking
- * @returns the highest ranking
- */
-function getHighestRanking<T>(item: T, keys: ReadonlyArray<KeyOption<T>>, value: string, options: Options<T>): RankingInfo {
-    const valuesToRank = getAllValuesToRank(item, keys);
-    return valuesToRank.reduce(
-        (acc, itemValue, index) => {
-            const ranking = acc;
-            const newRank = getMatchRanking(itemValue, value);
-            let newRankedValue = ranking.rankedValue;
-
-            if (newRank > ranking.rank) {
-                ranking.rank = newRank;
-                ranking.keyIndex = index;
-                ranking.keyThreshold = options.threshold;
-                newRankedValue = itemValue;
-            }
-            return {rankedValue: newRankedValue, rank: ranking.rank, keyIndex: ranking.keyIndex, keyThreshold: ranking.keyThreshold};
-        },
-        {
-            rankedValue: item as unknown as string,
-            rank: MATCH_RANK.NO_MATCH as Ranking,
-            keyIndex: -1,
-            keyThreshold: options.threshold,
-        },
-    );
-}
-
 /**
  * Takes an array of items and a value and returns a new array with the items that match the given value
  * @param items - the items to filter
  * @param searchValue - the value to use for ranking
- * @param options - options to configure
+ * @param extractRankableValuesFromItem - an array of functions
  * @returns the new filtered array
  */
-function filterArrayByMatch<T = string>(items: readonly T[], searchValue: string, options: Options<T>): T[] {
-    const {keys, threshold = MATCH_RANK.MATCHES} = options;
+function filterArrayByMatch<T = string>(items: readonly T[], searchValue: string, extractRankableValuesFromItem: (item: T) => string[]): T[] {
+    const filteredItems = [];
+    for (const item of items) {
+        const valuesToRank = extractRankableValuesFromItem(item);
+        let itemRank: Ranking = MATCH_RANK.NO_MATCH;
+        for (const value of valuesToRank) {
+            const rank = getMatchRanking(value, searchValue);
+            if (rank > itemRank) {
+                itemRank = rank;
+            }
+        }
 
-    return items
-        .map((item) => ({...getHighestRanking(item, keys, searchValue, options), item}))
-        .filter(({rank, keyThreshold = threshold}) => rank >= Math.max(keyThreshold, threshold + 1))
-        .map(({item}) => item);
+        if (itemRank >= MATCH_RANK.MATCHES) {
+            filteredItems.push(item);
+        }
+    }
+    return filteredItems;
 }
 
 export default filterArrayByMatch;
 export {MATCH_RANK};
-
-export type {Options, KeyAttributesOptions, KeyOption, RankingInfo, ValueGetterKey};

```

TMisiukiewicz committed:
Merge branch 'main' into perf/search-options-filtering

TMisiukiewicz committed:
update filtering method

TMisiukiewicz committed:
show chatrooms where user is a participant

TMisiukiewicz committed:
fix tests

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8732363773) to staging by https://github.com/roryabraham in version: 1.4.63-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8786104763) to production by https://github.com/mountiny in version: 1.4.63-21 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
